src/overlays/actors/ovl_Obj_Lightswitch/z_obj_lightswitch.c

```diff
@@ -27,47 +53,436 @@ const ActorInit Obj_Lightswitch_InitVars = {
+void ObjLightswitch_On(ObjLightswitch* this, GlobalContext* globalCtx) {
+    switch (this->actor.params >> 4 & 3) {
+        case OBJLIGHTSWITCH_TYPE_STAY_ON:
+            if (!Flags_GetSwitch(globalCtx, this->actor.params >> 8 & 0x3F)) {
+                ObjLightswitch_TurnOffInit(this);
+            }
+            break;
+        case OBJLIGHTSWITCH_TYPE_1:
+            if (this->collider.base.acFlags & 2 && !(this->prevFrameACflags & 2)) {
+                ObjLightswitch_TurnOffInit(this);
+                ObjLightswitch_ClearSwitchFlag(this, globalCtx);
+            }
+            break;
+        case OBJLIGHTSWITCH_TYPE_2:
+            if (!(this->collider.base.acFlags & 2)) {
+                if (this->timer >= 7) {
+                    ObjLightswitch_TurnOffInit(this);
+                    ObjLightswitch_ClearSwitchFlag(this, globalCtx);
+                } else {
+                    this->timer++;
+                }
+            } else {
+                this->timer = 0;
+            }
+            break;
+    }
+    this->flameRingRot += this->flameRingRotSpeed;
+}
+
+void ObjLightswitch_TurnOffInit(ObjLightswitch* this) {
+    this->actionFunc = ObjLightswitch_TurnOff;
+    this->toggleDelay = 100;
+    this->timer = 20;
+    this->faceTextureIndex = FACE_EYES_OPEN;
+}
+
+void ObjLightswitch_TurnOff(ObjLightswitch* this, GlobalContext* globalCtx) {
+    if ((this->actor.params >> 4 & 3) != OBJLIGHTSWITCH_TYPE_1 || func_8005B198() == this->actor.type ||
+        this->toggleDelay <= 0) {
+        this->timer--;
+
+        Math_ApproxS(&this->flameRingRotSpeed, 0, 0xA);
+        this->flameRingRot += this->flameRingRotSpeed;
+
+        this->color[0] = this->timer * (((255 - 155) << 6) / 20) + (155 << 6);
+        this->color[1] = this->timer * (((255 - 125) << 6) / 20) + (125 << 6);
+
+        if (this->timer <= 0) {
+            ObjLightswitch_OffInit(this);
+        } else if (this->timer == 15) {
+            this->faceTextureIndex = FACE_EYES_CLOSED;
+            Audio_PlayActorSound2(&this->actor, NA_SE_EV_FOOT_SWITCH);
+        }
+    }
+}
+
+void ObjLightswitch_DisappearDelayInit(ObjLightswitch* this) {
+    this->actionFunc = ObjLightswitch_DisappearDelay;
+    this->toggleDelay = 100;
+}
+
+void ObjLightswitch_DisappearDelay(ObjLightswitch* this, GlobalContext* globalCtx) {
+    if (func_8005B198() == this->actor.type || this->toggleDelay <= 0) {
+        ObjLightswitch_DisappearInit(this);
+    }
+}
+
+void ObjLightswitch_DisappearInit(ObjLightswitch* this) {
+    this->actionFunc = ObjLightswitch_Disappear;
+    this->alpha = 255 << 6;
+}
+
+void ObjLightswitch_Disappear(ObjLightswitch* this, GlobalContext* globalCtx) {
+    this->alpha -= 200;
+    ObjLightswitch_SpawnDisappearEffects(this, globalCtx);
+    if (this->alpha < 0) {
+        Actor_Kill(&this->actor);
+    }
+}
+
+void ObjLightswitch_Update(Actor* thisx, GlobalContext* globalCtx) {
+    ObjLightswitch* this = THIS;
+    Actor* child;
+
+    if (globalCtx) {}
+
```

Comment:
does globalCtx2 get rid of this by chance

Reply:
maybe it would if the stack allowed it but it doesn't, rip

Reply:
hmm, our theory is that its gonna be required for all main funcs that take thisx. so id hope it works with some more poking around

Reply:
child temp is def needed?

Reply:
and if its required, declaring it in the block its used possibly?

Reply:
nvm I can get rid of a temp and it does get rid of the if 👍

Reply:
perfect